consensus/types/src/beacon_state.rs

```diff
@@ -1669,6 +1671,59 @@ impl<T: EthSpec> BeaconState<T> {
         };
         Ok(sync_committee)
     }
+
+    pub fn compute_merkle_proof(
+        &mut self,
+        generalized_index: usize,
+    ) -> Result<Vec<Hash256>, Error> {
+        // 1. Convert generalized index to field index.
+        let field_index = match generalized_index {
+            light_client_update::CURRENT_SYNC_COMMITTEE_INDEX
+            | light_client_update::NEXT_SYNC_COMMITTEE_INDEX => {
+                // Sync committees are top-level fields, subtract off the generalized indices
+                // for the internal nodes. Result should be 22 or 23, the field offset of the committee
+                // in the `BeaconState`:
+                // https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#beaconstate
+                generalized_index.checked_sub(tree_hash_cache::NUM_BEACON_STATE_HASH_TREE_ROOT_LEAVES)
+                    .ok_or(Error::IndexNotSupported(generalized_index))?
+            }
+            light_client_update::FINALIZED_ROOT_INDEX => {
+                // Finalized root is the right child of `finalized_checkpoint`, divide by two to get
+                // the generalized index of `state.finalized_checkpoint`.
+                let finalized_checkpoint_generalized_index = generalized_index / 2;
+                // Subtract off the internal nodes. Result should be 105/2 - 32 = 20 which matches
+                // position of `finalized_checkpoint` in `BeaconState`.
+                finalized_checkpoint_generalized_index.checked_sub(tree_hash_cache::NUM_BEACON_STATE_HASH_TREE_ROOT_LEAVES)
+                    .ok_or(Error::IndexNotSupported(generalized_index))?
+            }
+            _ => return Err(Error::IndexNotSupported(generalized_index)),
+        };
+
+        // 2. Get all `BeaconState` leaves.
+        // The most efficient way to get these at the moment is from the tree hash cache.
+        // We can copy the guts of this function (or generalize it):
+        // https://github.com/sigp/lighthouse/blob/6d5a2b509fac7b6ffe693866f58ba49989f946d7/consensus/types/src/beacon_state/tree_hash_cache.rs#L213-L371
+        let cache = self.tree_hash_cache_mut().take();
+        let leaves;
+        if let Some(mut cache) = cache {
+            leaves = cache.recalculate_tree_hash_leaves(self)?;
+        } else {
+            return Err(Error::TreeHashCacheNotInitialized);
+        }
```

Comment:
the Rustiest way to write this is:

```suggestion
        let leaves = if let Some(mut cache) = cache {
            cache.recalculate_tree_hash_leaves(self)?
        } else {
            return Err(Error::TreeHashCacheNotInitialized);
        };
```

Reply:
aaaah my javascript web development class is corrupting me